### PR TITLE
DAOS-3658 tests: Create MACSio/hdf5 test

### PR DESCRIPTION
**Pull request by phender**
Create utility code for MACSio and use it to test hdf5.

Skip-func-hw: true
Test-tag: macsio

Signed-off-by: Phillip Henderson <phillip.henderson@intel.com>

### CODE REVIEW

File: src/tests/ftest/util/macsio_util.py [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+        #               Default depends on algorithm. By default, shuffling is
+        #               NOT used for zfp but IS used with all other algorithms.
+        #
+        #       Available compression algorithms:
+        #
+        #           "zfp"
+        #               Use Peter Lindstrom's ZFP compression (
+        #               computation.llnl.gov/casc/zfp) Note: Whether this
+        #               compression is available is determined entirely at
+        #               run-time using the H5Z-ZFP compresser as a generic
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/macsio_util.py]: Fixed

File: src/tests/ftest/util/macsio_util.py [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+        #       number of parts is this number multiplied by the MPI
+        #       communicator size. If the result of that product is
+        #       non-integral, it will be rounded and a warning message will be
+        #       generated.
+        self.avg_num_parts = FormattedParameter("--avg_num_parts {}")
+
+        #   --mesh_decomp %d %d %d []
+        #       The layout of parts in the mesh overriding the simple
+        #       decomposition e.g. 4 8 1 will decompose into 32 parts in the
+        #       stucture (x y z).
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/macsio_util.py]: Fixed

File: src/tests/ftest/util/macsio_test_base.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    def __init__(self, *args, **kwargs):
+        """Initialize a MacsioTestBase object."""
+        super(MacsioTestBase, self).__init__(*args, **kwargs)
+        self.manager = None
+        self.macsio = None
+
+    def setUp(self):
+        """Set up each test case."""
+        super(MacsioTestBase, self).setUp()
+        self.manager = Mpirun(None, subprocess=False, mpitype="mpich")
```
Comment (mjean308) [inline src/tests/ftest/util/macsio_test_base.py:45]: Do you need to consider the MPI env when hdf5 is built with openmpi?   Does Mpirun() use a full path ?  The concern here is what happens to LD_LIBRARY_PATH when you use Mpirun()?    If nothing then you should be ok to launch Macsio with mpich even when hdf5 is built with openmpi.

Comment (phender) [reply src/tests/ftest/util/macsio_test_base.py:45]: The `Mpirun` class handles setting the LD_LIBRARY_PATH with https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/job_manager_utils.py#L346

**Comment by phender** [comment]
The two failures in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-2106/20/ are known issues:

- job-2020-07-18T13.35-3b0382d-daos_test-daos_core_test (test_d): https://jira.hpdd.intel.com/browse/DAOS-5306
- job-2020-07-18T14.50-36ecb02-daos_test-daos_core_test-rebuild (test_r_26): https://jira.hpdd.intel.com/browse/DAOS-5331

File: src/tests/ftest/util/command_utils_base.py [hunk 4]
```diff
@@ -33,15 +33,25 @@ class CommandFailure(Exception):
 class BasicParameter(object):
     """A class for parameters whose values are read from a yaml file."""
 
-    def __init__(self, value, default=None):
+    def __init__(self, value, default=None, yaml_key=None):
         """Create a BasicParameter object.
 
+        Normal use includes assigning this object to a attribute name that
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/command_utils_base.py:39]: Applying this fix in https://github.com/daos-stack/daos/pull/3132. Thanks for catching it.

File: src/tests/ftest/util/command_utils_base.py [hunk 5]
```diff
@@ -33,15 +33,25 @@ class CommandFailure(Exception):
...
+        matches the test yaml file key used to assign its value.  If the
+        variable name will conflict with another class attribute, e.g. self.log,
+        then the `yaml_key` argument can used to define the test yaml file key
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/command_utils_base.py:42]: Applying this fix in https://github.com/daos-stack/daos/pull/3132. Thanks for catching it.

File: src/tests/ftest/util/command_utils_base.py [hunk 6]
```diff
@@ -102,51 +115,30 @@ def update_default(self, value):
...
-
 class FormattedParameter(BasicParameter):
     # pylint: disable=too-few-public-methods
     """A class for test parameters whose values are read from a yaml file."""
 
-    def __init__(self, str_format, default=None):
+    def __init__(self, str_format, default=None, yaml_key=None):
         """Create a FormattedParameter  object.
 
+        Normal use includes assigning this object to a attribute name that
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/command_utils_base.py:125]: Applying this fix in https://github.com/daos-stack/daos/pull/3132. Thanks for catching it.

File: src/tests/ftest/util/command_utils_base.py [hunk 7]
```diff
@@ -102,51 +115,30 @@ def update_default(self, value):
...
+        matches the test yaml file key used to assign its value.  If the
+        variable name will conflict with another class attribute, e.g. self.log,
+        then the `yaml_key` argument can used to define the test yaml file key
```
[the remark this answers is not in the document]
Comment (phender) [reply src/tests/ftest/util/command_utils_base.py:128]: Applying this fix in https://github.com/daos-stack/daos/pull/3132. Thanks for catching it.

File: Jenkinsfile [hunk 8]
```diff
@@ -121,7 +121,8 @@ el7_component_repos = ""
 def functional_rpms  = "--exclude openmpi openmpi3 hwloc ndctl " +
                        "ior-hpc-cart-4-daos-0 mpich-autoload-cart-4-daos-0 " +
                        "romio-tests-cart-4-daos-0 hdf5-tests-cart-4-daos-0 " +
-                       "mpi4py-tests-cart-4-daos-0 testmpio-cart-4-daos-0 fio"
+                       "mpi4py-tests-cart-4-daos-0 testmpio-cart-4-daos-0 " +
+                       "fio MACSio"
```
Comment (brianjmurrell) [inline Jenkinsfile:125]: Pity I didn't see this PR.  Adding test tool RPMs to `functional_rpms` should be deprecated and they should be added as dependencies of the _daos-tests_ RPM.  I just haven't gotten to making that change yet unfortunately to revert the existing ones.

**Pull request merged**